src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataComplexSerializer.cs

```diff
@@ -0,0 +1,123 @@
+        /// Initializes a new instance of the <see cref="ODataEdmTypeSerializer"/> class.
+        /// </summary>
+        /// <param name="payloadKind">The kind of OData payload that this serializer generates.</param>
+        /// <param name="serializerProvider">The <see cref="ODataSerializerProvider"/> to use to write inner objects.</param>
+        protected ODataComplexSerializer(ODataPayloadKind payloadKind, ODataSerializerProvider serializerProvider)
+            : base(payloadKind, serializerProvider)
+        {
+            if (serializerProvider == null)
+            {
+                throw Error.ArgumentNull("serializerProvider");
+            }
+          
+        }
+
+        /// <inheritdoc/>
+        public override void WriteObject(object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
+        {
+            if (messageWriter == null)
+            {
+                throw Error.ArgumentNull("messageWriter");
+            }
+            if (writeContext == null)
+            {
+                throw Error.ArgumentNull("writeContext");
+            }
+            if (writeContext.RootElementName == null)
+            {
+                throw Error.Argument("writeContext", SRResources.RootElementNameMissing, typeof(ODataSerializerContext).Name);
+            }
+
+            IEdmTypeReference edmType = writeContext.GetEdmType(graph, type);
+            Contract.Assert(edmType != null);
+
+            messageWriter.WriteProperty(CreateProperty(graph, edmType, writeContext.RootElementName, writeContext));
+        }
+
+        /// <inheritdoc/>
+        public sealed override ODataValue CreateODataValue(object graph, IEdmTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            if (!expectedType.IsComplex())
+            {
+                throw Error.InvalidOperation(SRResources.CannotWriteType, typeof(ODataComplexSerializer), expectedType.FullName());
+            }
+
+            ODataResourceValue value = CreateODataComplexResourceValue(graph, expectedType.AsComplex(), writeContext);
+            if (value == null)
+            {
+                return new ODataNullValue();
+            }
+
+            return value;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "edmTypeSerializer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        private ODataResourceValue CreateODataComplexResourceValue(object graph, IEdmComplexTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            ODataResourceValue resval = new ODataResourceValue { TypeName = expectedType.FullName() };
+            List<ODataProperty> lstProp = new List<ODataProperty>();
+
+            foreach (var prop in graph.GetType().GetProperties())
+            {
+                if (prop != null)
```

Comment:
>prop != null [](start = 20, length = 12)

when would prop be null?  Should this be an assert?

Reply:
for ignoring properties with null values

Reply:
but this doesn't ignore properties with null values, it ignores a null value for the propertyInfo, which should not occur.

---
In reply to: [460691078](https://github.com/OData/WebApi/pull/2219#discussion_r460691078) [](ancestors = 460691078)